storage: plumbing for DiskWriteCategory for categorized disk write metrics

CheranMahalingam opened the pull request:
This commit updates Pebble file creation calls to categorically track and display disk write metrics at the Cockroach layer.

Fixes https://github.com/cockroachdb/cockroach/issues/115434.

Epic: None.
Release note: None.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/119024)
<!-- Reviewable:end -->

CheranMahalingam commented:
New Disk Bandwidth Metrics

tpchvec/disk roachtest (Large data ingestion and SQL column spill):
<img width="1277" alt="Screenshot 2024-02-20 at 1 55 46 PM" src="https://github.com/cockroachdb/cockroach/assets/65926174/0a6dc2e9-902a-4070-a29b-85a7735b56df">

*Note that the sum of all disk write metrics does not perfectly match the collected OS write metrics, however, it follows a similar trend.

CheranMahalingam committed:
storage: plumbing for categorized disk write metrics

This commit updates Pebble file creation calls to categorically track and
display a breakdown of the source of writes to disk as a follow up to
https://github.com/cockroachdb/pebble/pull/3255.

Informs: #115434.

Epic: None.
Release note (general change): The following metrics were added for
observability into the source of disk writes:
- storage.category-pebble-wal.bytes-written
- storage.category-pebble-compaction.bytes-written
- storage.category-pebble-ingestion.bytes-written
- storage.category-pebble-memtable-flush.bytes-written
- storage.category-raft-snapshot.bytes-written
- storage.category-encryption-registry.bytes-written
- storage.category-crdb-log.bytes-written
- storage.category-sql-row-spill.bytes-written
- storage.category-sql-col-spill.bytes-written
- storage.category-unspecified.bytes-written

CheranMahalingam committed:
ui: add disk write breakdown to storage dashboard

Fixes: #115434.

Epic: None.
Release note (ui change): The Storage metrics dashboard now contains a
graph categorizing disk writes that contribute to
`sys.host.disk.write.bytes` according to the source of the write (WAL,
compactions, sstable ingestion, memtable flushes, raft snapshots,
encryption registry, logs, SQL columnar spill, or SQL row spill).

aadityasondhi reviewed (changes requested):
Great work!  Mostly just nits and a couple of questions here.

> Reviewed 10 of 76 files at r1, 105 of 122 files at r2, 34 of 34 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-:-Nu_CH0z8apvAPPbzi8X:b5krjlc)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @CheranMahalingam, @DrewKimball, @herkolategan, @jbowens, @renatolabs, and @xinhaoz)

___
*[`pkg/cli/debug.go` line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_4wugCAIBhel_EJ12:-Nu_4wugCAIBhel_EJ13:b-tfgwd) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cli/debug.go#L187)):*
> ```Go
> 		}
> 	}
> 	return fs.InitEnv(context.Background(), vfs.Default, dir, envConfig, nil)
> ```

nit: can we add a comment to identify what the nil is? ex. `nil /* statsCollector */`
___
*[`pkg/cli/debug.go` line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_6GhEEpXIfZYdLUX7:-Nu_6GhEEpXIfZYdLUX8:bnaxzq5) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cli/debug.go#L187)):*
> ```Go
> 		}
> 	}
> 	return fs.InitEnv(context.Background(), vfs.Default, dir, envConfig, nil)
> ```

On a separate note, why do we not have a stat collector here?
___
*[`pkg/cli/debug.go` line 1646 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_5LHK7ZOGEqBaFKaS:-Nu_5LHK7ZOGEqBaFKaT:bknz1y4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cli/debug.go#L1646)):*
> ```Go
> 				return nil, err
> 			}
> 			env, err := fs.InitEnv(ctx, vfs.Default, dir, envConfig, nil)
> ```

nit: comment on the nil
___
*[`pkg/cmd/roachtest/run.go` line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_79ju7XEut9W5CoO8:-Nu_79ju7XEut9W5CoO9:bh17cmi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cmd/roachtest/run.go#L208)):*
> ```Go
> 		panic(err)
> 	}
> 	if _, err := log.ApplyConfig(logConf, log.FileSinkMetrics{}); err != nil {
> ```

Why do we pass an empty `FileSinkMetrics` here?
___
*[`pkg/storage/metamorphic/generator.go` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_AL3OCuHtk4L2Awoc:-Nu_AL3OCuHtk4L2Awod:be5akbz) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/metamorphic/generator.go#L44)):*
> ```Go
> 
> func (e *engineConfig) create(path string, baseFS vfs.FS) (storage.Engine, error) {
> 	env, err := fs.InitEnv(context.Background(), baseFS, path, fs.EnvConfig{}, nil)
> ```

comment on what the nil variable is
___
*[`pkg/kv/kvclient/kvstreamer/results_buffer_test.go` line 47 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_ApYN66MZon9jO81T:-Nu_ApYN66MZon9jO81U:b9devkq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvclient/kvstreamer/results_buffer_test.go#L47)):*
> ```Go
> 		base.DefaultTestTempStorageConfig(st),
> 		base.DefaultTestStoreSpec,
> 		nil,
> ```

comment what the nil variable is
___
*[`pkg/kv/kvserver/client_metrics_test.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Avf16f_N-Wda--xS:-Nu_Avf2CWXuUxEO84af:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/client_metrics_test.go#L107)):*
> ```Go
> 		// Engine again in read-only mode (leaving the rest of the
> 		// Server stopped) to compute MVCC stats.
> 		env, err := fs.InitEnvFromStoreSpec(ctx, specs[storeIdx], fs.ReadOnly, stickyRegistry, nil)
> ```

same as above
___
*[`pkg/kv/kvserver/consistency_queue_test.go` line 393 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_B3Mo6ThXpwt8AHwT:-Nu_B3Mo6ThXpwt8AHwU:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/consistency_queue_test.go#L393)):*
> ```Go
> 		ctx := context.Background()
> 		memFS := stickyVFSRegistry.Get(strconv.FormatInt(int64(i), 10))
> 		env, err := fs.InitEnv(ctx, memFS, cps[0], fs.EnvConfig{RW: fs.ReadOnly}, nil)
> ```

same as above
___
*[`pkg/kv/kvserver/batcheval/cmd_refresh_range_bench_test.go` line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_B8oK9VE7d22SNppY:-Nu_B8oK9VE7d22SNppZ:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/batcheval/cmd_refresh_range_bench_test.go#L186)):*
> ```Go
> 
> func setupMVCCPebble(b testing.TB, dir string, lBaseMaxBytes int64, rw fs.RWMode) storage.Engine {
> 	env, err := fs.InitEnv(context.Background(), vfs.Default, dir, fs.EnvConfig{RW: rw}, nil)
> ```

same as above
___
*[`pkg/kv/kvserver/logstore/sideload_test.go` line 611 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BBl-5zIDRvxZIKk6:-Nu_BBl-5zIDRvxZIKk7:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/logstore/sideload_test.go#L611)):*
> ```Go
> 		ctx := context.Background()
> 		memFS := vfs.NewStrictMem()
> 		env, err := fs.InitEnv(ctx, memFS, "", fs.EnvConfig{}, nil)
> ```

same as above
___
*[`pkg/kv/kvserver/logstore/sideload_test.go` line 633 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BCwk0mQaxJq0orh1:-Nu_BCwk0mQaxJq0orh2:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/logstore/sideload_test.go#L633)):*
> ```Go
> 
> 		// Emulate process restart. Load from the last synced state.
> 		env, err = fs.InitEnv(ctx, memFS, "", fs.EnvConfig{}, nil)
> ```

same as above
___
*[`pkg/kv/kvserver/rangefeed/catchup_scan_bench_test.go` line 207 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BG38-Zj3osJpVarC:-Nu_BG38-Zj3osJpVarD:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/rangefeed/catchup_scan_bench_test.go#L207)):*
> ```Go
> 
> func setupMVCCPebble(b testing.TB, dir string, lBaseMaxBytes int64, rw fs.RWMode) storage.Engine {
> 	env, err := fs.InitEnv(context.Background(), vfs.Default, dir, fs.EnvConfig{RW: rw}, nil)
> ```

same as above
___
*[`pkg/sql/distsql/columnar_utils_test.go` line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BNfVDzJwVY32nScA:-Nu_BNfVDzJwVY32nScB:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/distsql/columnar_utils_test.go#L90)):*
> ```Go
> 	ctx := context.Background()
> 	st := cluster.MakeTestingClusterSettings()
> 	tempEngine, tempFS, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowcontainer/disk_row_container_test.go` line 126 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BRBaAqs1RbF9rDWb:-Nu_BRBaAqs1RbF9rDWc:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/disk_row_container_test.go#L126)):*
> ```Go
> 	ctx := context.Background()
> 	st := cluster.MakeTestingClusterSettings()
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowcontainer/hash_row_container_test.go` line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BUiF98sqUdI-OUVq:-Nu_BUiF98sqUdI-OUVr:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/hash_row_container_test.go#L42)):*
> ```Go
> 	st := cluster.MakeTestingClusterSettings()
> 	evalCtx := eval.MakeTestingEvalContext(st)
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowcontainer/numbered_row_container_test.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BWdFB5AFsvHW7iNN:-Nu_BWdFB5AFsvHW7iNO:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/numbered_row_container_test.go#L52)):*
> ```Go
> 	st := cluster.MakeTestingClusterSettings()
> 	evalCtx := eval.MakeTestingEvalContext(st)
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowcontainer/row_container_test.go` line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BZ4rBUSZEF4BrtAb:-Nu_BZ4rBUSZEF4BrtAc:binws3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/row_container_test.go#L212)):*
> ```Go
> 		},
> 		base.DefaultTestStoreSpec,
> 		nil,
> ```

comment here the variable name
___
*[`pkg/sql/rowexec/hashjoiner_test.go` line 1025 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BaxFCjHUBtyCNr4O:-Nu_BaxFCjHUBtyCNr4P:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/hashjoiner_test.go#L1025)):*
> ```Go
> 	ctx := context.Background()
> 	st := cluster.MakeTestingClusterSettings()
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowexec/inverted_joiner_test.go` line 633 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Bcvn0lleoW27S0l2:-Nu_Bcvn0lleoW27S0l3:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/inverted_joiner_test.go#L633)):*
> ```Go
> 
> 	st := cluster.MakeTestingClusterSettings()
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowexec/joinreader_test.go` line 1058 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Beqx5J8lAurMkDBB:-Nu_Beqx5J8lAurMkDBC:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/joinreader_test.go#L1058)):*
> ```Go
> 	}
> 	st := cluster.MakeTestingClusterSettings()
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowexec/sorter_test.go` line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BhEt9k1WN34FWkYm:-Nu_BhEt9k1WN34FWkYn:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/sorter_test.go#L269)):*
> ```Go
> 					ctx := context.Background()
> 					st := cluster.MakeTestingClusterSettings()
> 					tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowexec/windower_test.go` line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BjRY2ON1EqQV4I4Y:-Nu_BjRY2ON1EqQV4I4Z:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/windower_test.go#L53)):*
> ```Go
> 	diskMonitor := execinfra.NewTestDiskMonitor(ctx, st)
> 	defer diskMonitor.Stop(ctx)
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/sql/rowflow/routers_test.go` line 767 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BleJ4IR4p7YWVrSg:-Nu_BleJ4IR4p7YWVrSh:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowflow/routers_test.go#L767)):*
> ```Go
> 	diskMonitor := execinfra.NewTestDiskMonitor(ctx, st)
> 	defer diskMonitor.Stop(ctx)
> 	tempEngine, _, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec, nil)
> ```

same as above
___
*[`pkg/storage/bench_data_test.go` line 135 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BoH4C5lksqA8mwmF:-Nu_BoH4C5lksqA8mwmG:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/bench_data_test.go#L135)):*
> ```Go
> 		// performance.
> 		buildFS = vfs.NewMem()
> 		env, err := fs.InitEnv(ctx, buildFS, "", fs.EnvConfig{}, nil)
> ```

same as above
___
*[`pkg/storage/disk_map_test.go` line 183 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BrX31x6bGDE530Ju:-Nu_BrX31x6bGDE530Jv:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/disk_map_test.go#L183)):*
> ```Go
> 		Path:     dir,
> 		Settings: cluster.MakeClusterSettings(),
> 	}, base.StoreSpec{}, nil)
> ```

same as above
___
*[`pkg/storage/min_version_test.go` line 130 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Bv7vBTyx7lt7WyCf:-Nu_Bv7vBTyx7lt7WyCg:bnj7hl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/min_version_test.go#L130)):*
> ```Go
> 	env, err := fs.InitEnv(ctx, baseFS, "", fs.EnvConfig{
> 		EncryptionOptions: []byte("foo"),
> 	}, nil)
> ```

same as above


<!-- Sent from Reviewable.io -->

CheranMahalingam reviewed (commented):
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-:-Nu_dJUb3CgGPZOvRzZ2:b-814g8k)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @DrewKimball, @herkolategan, @jbowens, @renatolabs, and @xinhaoz)

___
*[`DEPS.bzl` line 1690 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Ntc8ZuYD-f0MWOmK36r:-Ntc8ZuYD-f0MWOmK36s:b-4visa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/dce01742ae64912eb971ff7491344490ba621d3e/DEPS.bzl#L1690)):*
> ```Starlark
>         importpath = "github.com/cockroachdb/pebble",
>         vcs = "git",
>         remote = "https://github.com/CheranMahalingam/pebble",
> ```

Modified in order to get CI to build and run tests, however, this should be reverted once this PR has been reviewed and https://github.com/cockroachdb/pebble/pull/3255 has been merged. Note that for whatever reason, the linter still doesn't seem to pick up on the vfs interface changes.
___
*[`pkg/cli/debug.go` line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_6GhEEpXIfZYdLUX7:-Nu_UUy4BgCGjpL6x03-:bhga057) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cli/debug.go#L187)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

On a separate note, why do we not have a stat collector here?
</blockquote></details>

I am not very familiar with this part of the codebase, but I assumed these function were invoked as part of the `cockroach debug` commands for dumping logs in a zip or tsdump. If that is the case, I don't know if there is a benefit to tracking disk writes since this is running separately from the cockroach node. Also, they seem to use the returned filesystem to open a Pebble engine and most of the time it is opened in read-only mode so it won't contribute to Pebble writes (exceptions are for recovering ranges from quorum loss and synctest which is no longer used).
___
*[`pkg/cmd/roachtest/run.go` line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_79ju7XEut9W5CoO8:-NudPcez94CTnTOrwr06:b8hf4ax) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cmd/roachtest/run.go#L208)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

Why do we pass an empty `FileSinkMetrics` here?
</blockquote></details>

This method seems to be invoked by the roachtest runner which is separate from the cluster that gets set up remotely. Each node in the cluster will go through the normal setup which will include the code paths for setting up logging metrics. I can also confirm that I get the new crdb-log disk write metrics when I run roachtests.
___
*[`pkg/storage/metamorphic/generator.go` line 44 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_AL3OCuHtk4L2Awoc:-Nu_ZjnM84X_0-hziFAH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/metamorphic/generator.go#L44)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

comment on what the nil variable is
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvstreamer/results_buffer_test.go` line 47 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_ApYN66MZon9jO81T:-Nu_ZyPT2_UUAyYalptg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvclient/kvstreamer/results_buffer_test.go#L47)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

comment what the nil variable is
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/client_metrics_test.go` line 107 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Avf16f_N-Wda--xS:-Nu__AcX87XjHmmyo4Xp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/client_metrics_test.go#L107)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/consistency_queue_test.go` line 393 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_B3Mo6ThXpwt8AHwT:-Nu__HUR5pYnBOPAxHqJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/consistency_queue_test.go#L393)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/batcheval/cmd_refresh_range_bench_test.go` line 186 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_B8oK9VE7d22SNppY:-Nu__RLh9Ul7iyUgBONG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/batcheval/cmd_refresh_range_bench_test.go#L186)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/logstore/sideload_test.go` line 611 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BBl-5zIDRvxZIKk6:-Nu__aNy4FUuk2R6lAFS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/logstore/sideload_test.go#L611)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/logstore/sideload_test.go` line 633 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BCwk0mQaxJq0orh1:-Nu__hMp4ETVCVf591Tl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/logstore/sideload_test.go#L633)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/rangefeed/catchup_scan_bench_test.go` line 207 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BG38-Zj3osJpVarC:-Nu__qk78_C5TAGJX-GX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/kv/kvserver/rangefeed/catchup_scan_bench_test.go#L207)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/distsql/columnar_utils_test.go` line 90 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BNfVDzJwVY32nScA:-Nu__yr916SrwAedzMSw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/distsql/columnar_utils_test.go#L90)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowcontainer/disk_row_container_test.go` line 126 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BRBaAqs1RbF9rDWb:-Nu_aKp36eIrq-m-IHHX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/disk_row_container_test.go#L126)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowcontainer/hash_row_container_test.go` line 42 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BUiF98sqUdI-OUVq:-Nu_aSi0-NYuePfvh9H5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/hash_row_container_test.go#L42)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowcontainer/numbered_row_container_test.go` line 52 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BWdFB5AFsvHW7iNN:-Nu_asd43xd67KrLx0YL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/numbered_row_container_test.go#L52)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowcontainer/row_container_test.go` line 212 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BZ4rBUSZEF4BrtAb:-Nu_bJ_c9OVByYUl48bF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowcontainer/row_container_test.go#L212)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

comment here the variable name
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/hashjoiner_test.go` line 1025 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BaxFCjHUBtyCNr4O:-Nu_bf3J5NG2-ZqpNmjb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/hashjoiner_test.go#L1025)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/inverted_joiner_test.go` line 633 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Bcvn0lleoW27S0l2:-Nu_bvJBBlQ0s_9v7Cmj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/inverted_joiner_test.go#L633)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/joinreader_test.go` line 1058 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Beqx5J8lAurMkDBB:-Nu_cFOS2Vho8kbbJXRX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/joinreader_test.go#L1058)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/sorter_test.go` line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BhEt9k1WN34FWkYm:-Nu_cMAsC5xCBIreYscW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/sorter_test.go#L269)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowexec/windower_test.go` line 53 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BjRY2ON1EqQV4I4Y:-Nu_cS7a1kLiXbR8Ol86:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowexec/windower_test.go#L53)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/sql/rowflow/routers_test.go` line 767 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BleJ4IR4p7YWVrSg:-Nu_c_Rf6iCxpd_opKRy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/sql/rowflow/routers_test.go#L767)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/storage/bench_data_test.go` line 135 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BoH4C5lksqA8mwmF:-Nu_ckW_BrI5e_K8nBOI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/bench_data_test.go#L135)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/storage/disk_map_test.go` line 183 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_BrX31x6bGDE530Ju:-Nu_d2Tq9GG3D9JWL0c_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/disk_map_test.go#L183)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.
___
*[`pkg/storage/min_version_test.go` line 130 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_Bv7vBTyx7lt7WyCf:-Nu_dBNA73A8aPXmxi-K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/storage/min_version_test.go#L130)):*
<details><summary><i>Previously, aadityasondhi (Aaditya Sondhi) wrote…</i></summary><blockquote>

same as above
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

CheranMahalingam reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119024)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi, @DrewKimball, @herkolategan, @jbowens, @renatolabs, and @xinhaoz)

___
*[`pkg/cmd/roachtest/run.go` line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_79ju7XEut9W5CoO8:-NudVsia8cqWY4rmwaN_:b-2q8ieh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cmd/roachtest/run.go#L208)):*
<details><summary><i>Previously, CheranMahalingam wrote…</i></summary><blockquote>

This method seems to be invoked by the roachtest runner which is separate from the cluster that gets set up remotely. Each node in the cluster will go through the normal setup which will include the code paths for setting up logging metrics. I can also confirm that I get the new crdb-log disk write metrics when I run roachtests.
</blockquote></details>

![Screenshot 2024-04-04 at 10.28.40 AM.png](https://files.reviewableusercontent.io/images/reviewable/65926174/Cz1efDv3MtpKoyBjvBXe0Q2r/Screenshot%202024-04-04%20at%2010.28.40%E2%80%AFAM.png)


<!-- Sent from Reviewable.io -->

aadityasondhi reviewed (approved):
Great work on this!

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-:-NudX97539i5CCz_lbUG:b4u7uhx)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @CheranMahalingam, @DrewKimball, @herkolategan, @jbowens, @renatolabs, and @xinhaoz)

___
*[`pkg/cli/debug.go` line 187 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_6GhEEpXIfZYdLUX7:-NudWxblDsX3uOgREfIw:b-7sy5bg) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cli/debug.go#L187)):*
<details><summary><i>Previously, CheranMahalingam wrote…</i></summary><blockquote>

I am not very familiar with this part of the codebase, but I assumed these function were invoked as part of the `cockroach debug` commands for dumping logs in a zip or tsdump. If that is the case, I don't know if there is a benefit to tracking disk writes since this is running separately from the cockroach node. Also, they seem to use the returned filesystem to open a Pebble engine and most of the time it is opened in read-only mode so it won't contribute to Pebble writes (exceptions are for recovering ranges from quorum loss and synctest which is no longer used).
</blockquote></details>

Yeah that makes sense. 
___
*[`pkg/cmd/roachtest/run.go` line 208 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119024#-Nu_79ju7XEut9W5CoO8:-NudX3bR5xDRAUDafY4u:b6i4gr3) ([raw file](https://github.com/cockroachdb/cockroach/blob/a380e6070d79200e1d98bec44b1e9214cabb38ad/pkg/cmd/roachtest/run.go#L208)):*
<details><summary><i>Previously, CheranMahalingam wrote…</i></summary><blockquote>

![Screenshot 2024-04-04 at 10.28.40 AM.png](https://files.reviewableusercontent.io/images/reviewable/65926174/Cz1efDv3MtpKoyBjvBXe0Q2r/Screenshot%202024-04-04%20at%2010.28.40%E2%80%AFAM.png)
</blockquote></details>

Great thanks for testing it out.


<!-- Sent from Reviewable.io -->

CheranMahalingam committed:
go.mod: bump Pebble to 278b6a67a6f0

Changes:

 * [`278b6a67`](https://github.com/cockroachdb/pebble/commit/278b6a67) vfs: add write category during file creation
 * [`45b7a800`](https://github.com/cockroachdb/pebble/commit/45b7a800) sstable: strengthen assertion around mutual exclusion of seek optimizations
 * [`56554864`](https://github.com/cockroachdb/pebble/commit/56554864) sstable: remove Iterator.MaybeFilteredKeys
 * [`f31f2611`](https://github.com/cockroachdb/pebble/commit/f31f2611) db: unconditionally interleave ignorable boundaries

Release note: none.
Epic: none.

CheranMahalingam commented:
bors r=aadityasondhi
